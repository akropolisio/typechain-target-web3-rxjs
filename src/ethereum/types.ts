import { EventLog as Web3EventLog, TransactionReceipt } from 'web3-core';
import BN from 'bn.js';

export type EventLog<T> = Omit<Web3EventLog, 'returnValues'> & { returnValues: T };

export interface EventEmitter<T> {
  on(type: 'data', handler: (event: EventLog<T>) => void): EventEmitter<T>;
  on(type: 'changed', handler: (receipt: EventLog<T>) => void): EventEmitter<T>;
  on(type: 'error', handler: (error: Error) => void): EventEmitter<T>;
  on(
    type: 'error' | 'data' | 'changed',
    handler: (error: Error | TransactionReceipt | string) => void,
  ): EventEmitter<T>;
}

export interface Tx {
  nonce?: string | number;
  chainId?: string | number;
  from?: string;
  to?: string;
  data?: string;
  value?: string | number;
  gas?: string | number;
  gasPrice?: string | number;
  maxPriorityFeePerGas?: string | number;
  maxFeePerGas?: string | number;
}

export type InputEvmTypeToJSTypeMap = {
  address: string;
  integer: BN;
  uinteger: BN;
  boolean: boolean;
  string: string;
  bytes: string;
  'dynamic-bytes': string;
};

export type InputEvmType = keyof InputEvmTypeToJSTypeMap;

export type JSType = InputEvmTypeToJSTypeMap[InputEvmType];
