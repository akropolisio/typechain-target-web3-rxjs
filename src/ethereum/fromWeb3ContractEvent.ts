import { fromEventPattern, Observable } from 'rxjs';
import { EventData } from 'web3-eth-contract';

type EventEmitter = {
  on(event: string, handler: (event: EventData) => void): EventEmitter;
};

export function fromWeb3DataEvent(emitter: EventEmitter): Observable<EventData> {
  interface IUnsubscribable {
    unsubscribe: () => void;
  }

  return fromEventPattern<EventData>(
    handler => emitter.on('data', handler),
    (_, signal: IUnsubscribable) => signal.unsubscribe(),
  );
}
