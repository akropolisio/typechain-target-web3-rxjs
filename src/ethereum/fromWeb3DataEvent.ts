import { fromEventPattern, Observable } from 'rxjs';
import { EventLog } from 'web3-core';

import { EventEmitter } from './types';

export function fromWeb3DataEvent(emitter: EventEmitter<any>): Observable<EventLog> {
  interface IUnsubscribable {
    unsubscribe: () => void;
  }

  return fromEventPattern<EventLog>(
    handler => emitter.on('data', handler as (event: EventLog) => void),
    (_, signal: IUnsubscribable) => signal.unsubscribe(),
  );
}
