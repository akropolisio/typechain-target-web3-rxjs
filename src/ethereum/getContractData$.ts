import { Observable, from, merge, EMPTY, ReplaySubject } from 'rxjs';
import { skipUntil, mergeMap, throttleTime, switchMap, shareReplay } from 'rxjs/operators';
import { Eth } from 'web3-eth';
import { Contract } from 'web3-eth-contract';

import {
  EventLog,
  CallOptions,
  Web3ContractInputArgs,
  Web3ContractResponse,
  Response,
} from './types';

interface IOptions {
  eventsForReload?: Observable<EventLog<any>> | Observable<EventLog<any>>[];
  reloadTrigger$?: Observable<any>;
  args?: Web3ContractInputArgs;
  convert?(value: Web3ContractResponse): Response | void;
  tx?: CallOptions;
}

function identity(value: any) {
  return value;
}

function awaitMs(delayMs: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, delayMs));
}

export function getContractData$(
  contract: Contract,
  eth: Eth,
  method: string,
  options: IOptions = {},
): Observable<Response> {
  const { eventsForReload = [], reloadTrigger$ = EMPTY, args = [], convert = identity } = options;

  const load = async () => {
    const data = await contract.methods[method](...args).call(options.tx);
    return convert(data);
  };

  const reloadEvents = Array.isArray(eventsForReload) ? eventsForReload : [eventsForReload];

  const first$ = from(load());
  const fromEvents$ = merge(...reloadEvents).pipe(
    skipUntil(first$),
    throttleTime(200),
    switchMap(async event => {
      let currentBlock = await eth.getBlockNumber();

      /* eslint-disable no-await-in-loop */
      while (currentBlock < event.blockNumber) {
        await awaitMs(500);
        currentBlock = await eth.getBlockNumber();
      }
      /* eslint-enable no-await-in-loop */

      return event;
    }),
    mergeMap(() => from(load()), 1),
    shareReplay(1),
  );

  const subject = new ReplaySubject<Response>(1);

  merge(first$, fromEvents$, reloadTrigger$).subscribe(subject);

  return subject;
}
