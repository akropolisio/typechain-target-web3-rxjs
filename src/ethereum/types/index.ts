export * from './core';
export * from './contractWrapper';
