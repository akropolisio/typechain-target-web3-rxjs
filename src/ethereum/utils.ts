import { RawAbiDefinition } from 'typechain';
// @ts-ignore
import { _jsonInterfaceMethodToString, AbiItem } from 'web3-utils';

export function extractAbiMethod(abi: RawAbiDefinition[], method: string): RawAbiDefinition | null {
  const [, methodName] = method.match(/(\w+?)($|\(.*)/) || [];

  if (!methodName) {
    console.warn(new Error(`Invalid method: ${method}`));
    return null;
  }

  const methodEntries = abi.filter(m => m.name === methodName && m.type === 'function');

  if (!methodEntries.length) {
    console.warn(new Error(`Method ${method} not found`));
    return null;
  }

  const isSimpleMethodName = method === methodName;
  if (isSimpleMethodName) {
    return methodEntries[0];
  }

  const entry = methodEntries.find(m => _jsonInterfaceMethodToString(m as AbiItem) === method);

  if (!entry) {
    console.warn(new Error(`Method ${method} not found`));
    return null;
  }

  return entry;
}

export function attachStaticFields<T extends {}, I extends Record<string, any>>(
  target: T,
  staticFields: I,
): T & I {
  const result: T & I = target as T & I;

  Object.keys(staticFields).forEach((key: keyof I) => {
    (result as I)[key] = staticFields[key];
  });

  return result;
}
