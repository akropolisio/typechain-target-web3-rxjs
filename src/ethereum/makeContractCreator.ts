import { Observable } from 'rxjs';
import { A, B } from 'ts-toolbelt';
import BN from 'bn.js';
import Web3 from 'web3';
import { PromiEvent, TransactionReceipt } from 'web3-core';
import { Contract } from 'web3-eth-contract';

import { getContractData$ } from './getContractData$';
import {
  EventEmitter,
  SendOptions,
  CallOptions,
  TransactionObject,
  EventLog,
  InputEvmType,
  InputEvmTypeToJSTypeMap,
  JSType,
} from './types';

/* ***** OVERRIDE WEB3 TYPES ***** */

type BlockType = 'latest' | 'pending' | 'genesis' | number;

type Callback<T> = (error: Error, result: T) => void;

/* ***** */

interface GenericDescriptor {
  callMethods: Record<string, MethodDescriptor>;
  sendMethods: Record<string, MethodDescriptor>;
  events: Record<string, EventDescriptor>;
}

type MethodDescriptor = {
  inputs?: readonly NamedInput<string, InputEvmType | Input<InputEvmType, boolean>, boolean>[];
  output?: readonly Output<OutputEvmType, boolean>[];
};

interface EventDescriptor {
  inputs: readonly NamedInput<string, InputEvmType | Input<InputEvmType, boolean>, boolean>[];
}

interface NamedInput<
  N extends string = string,
  T extends InputEvmType | Input<InputEvmType, boolean> =
    | InputEvmType
    | Input<InputEvmType, boolean>,
  IA extends boolean = false
> {
  name: N;
  type: T;
  isArray: IA;
}

interface Input<T extends InputEvmType = InputEvmType, IA extends boolean = false> {
  type: T;
  isArray: IA;
}

type Output<T extends OutputEvmType = OutputEvmType, IA extends boolean = false> = {
  type: T;
  isArray: IA;
};

type OutputEvmTypeToJSTypeMap = InputEvmTypeToJSTypeMap & {
  void: void;
};
type OutputEvmType = keyof OutputEvmTypeToJSTypeMap;

type IOToJSType<
  T extends Output<any, any>,
  ToJSTypeMap extends Record<string, JSType | void>
> = T['isArray'] extends true ? ToJSTypeMap[T['type']][] : ToJSTypeMap[T['type']];

type InferTypeProp<T> = T extends NamedInput<infer Name, infer Type, boolean>
  ? Type extends OutputEvmType
    ? {
        [key in Name]: IOToJSType<T, InputEvmTypeToJSTypeMap>;
      }
    : { [key in Name]: IOToJSType<Exclude<Type, OutputEvmType>, InputEvmTypeToJSTypeMap>[] }
  : never;
type InputsToArg<T> = MergeTupleMembers<{ [P in keyof T]: InferTypeProp<T[P]> }>;
type MaybeInputsToArgs<S> = S extends readonly any[]
  ? A.Equals<S, readonly []> extends B.True
    ? void
    : A.Equals<S, []> extends B.True
    ? void
    : InputsToArg<S>
  : void;

type MaybeOutputToResponse<O> = O extends readonly any[]
  ? A.Equals<O, readonly []> extends B.True
    ? void
    : A.Equals<O, []> extends B.True
    ? void
    : O extends [any] | readonly [any]
    ? IOToJSType<Extract<O[0], Output<OutputEvmType, boolean>>, OutputEvmTypeToJSTypeMap>
    : {
        -readonly [key in keyof O]: IOToJSType<Extract<O[key], Output>, OutputEvmTypeToJSTypeMap>;
      }
  : void;

type CallMethod<M extends MethodDescriptor = MethodDescriptor> = (
  input: MaybeInputsToArgs<M['inputs']>,
  eventsForReload?: EventEmitter<any> | EventEmitter<any>[],
  updatingDelay?: number,
  tx?: CallOptions,
) => Observable<MaybeOutputToResponse<NonNullable<M['output']>>>;

// for read non view methods
type ReadMethod<M extends MethodDescriptor = MethodDescriptor> = (
  input: MaybeInputsToArgs<M['inputs']>,
  tx: SendOptions,
  eventsForReload?: EventEmitter<any> | EventEmitter<any>[],
  updatingDelay?: number,
) => Observable<MaybeOutputToResponse<NonNullable<M['output']>>>;

type SendMethod<M extends MethodDescriptor = MethodDescriptor> = ((
  input: MaybeInputsToArgs<M['inputs']>,
  tx: SendOptions,
) => PromiEvent<TransactionReceipt>) & {
  getTransaction(input: MaybeInputsToArgs<M['inputs']>): TransactionObject;
  read: ReadMethod<M>;
};

type EventMethod<E extends EventDescriptor> = (
  options?: SubscribeEventOptions<E>,
  cb?: Callback<EventLog<MaybeInputsToArgs<E>>>,
) => EventEmitter<MaybeInputsToArgs<E>>;

interface SubscribeEventOptions<E extends EventDescriptor> {
  filter?: Partial<MaybeInputsToArgs<E['inputs']>>;
  fromBlock?: BlockType;
  topics?: string[];
}

type ContractWrapper<D extends GenericDescriptor> = {
  methods: {
    [key in keyof D['callMethods']]: CallMethod<D['callMethods'][key]>;
  } &
    {
      [key in keyof D['sendMethods']]: SendMethod<D['sendMethods'][key]>;
    };
  events: {
    [key in keyof D['events']]: EventMethod<D['events'][key]>;
  } & {
    allEvents: Contract['events']['allEvents'];
  };
  getPastEvents: Contract['getPastEvents'];
};

export function getNamedInput<
  N extends string,
  T extends InputEvmType | Input<InputEvmType, boolean>
>(name: N, type: T): NamedInput<N, T>;
export function getNamedInput<
  N extends string,
  T extends InputEvmType | Input<InputEvmType, boolean>
>(name: N, type: T, isArray: true): NamedInput<N, T, true>;
export function getNamedInput<
  N extends string,
  T extends InputEvmType | Input<InputEvmType, boolean>
>(name: N, type: T, isArray: boolean = false): NamedInput<N, T, boolean> {
  return { name, type, isArray };
}

export function getInput<T extends InputEvmType>(type: T): Input<T>;
export function getInput<T extends InputEvmType>(type: T, isArray: true): Input<T, true>;
export function getInput<T extends InputEvmType>(
  type: T,
  isArray: boolean = false,
): Input<T, boolean> {
  return { type, isArray };
}

export function getOutput<T extends OutputEvmType = OutputEvmType>(type: T): Output<T>;
export function getOutput<T extends OutputEvmType = OutputEvmType>(
  type: T,
  isArray: true,
): Output<T, true>;
export function getOutput<T extends OutputEvmType = OutputEvmType>(
  type: T,
  isArray: boolean = false,
): Output<T, boolean> {
  return { type, isArray };
}

const toRequest: {
  [key in InputEvmType]: (input: InputEvmTypeToJSTypeMap[key]) => JSType;
} = {
  address: value => value,
  boolean: value => value,
  integer: value => value.toString(),
  uinteger: value => value.toString(),
  string: value => value,
  bytes: value => value,
  'dynamic-bytes': value => value,
};

const fromResponse: {
  [key in OutputEvmType]: (input: JSType) => OutputEvmTypeToJSTypeMap[key];
} = {
  address: value => String(value),
  boolean: value => Boolean(value),
  integer: value => new BN(value as string | BN),
  uinteger: value => new BN(value as string | BN),
  string: value => String(value),
  bytes: value => String(value),
  'dynamic-bytes': value => String(value),
  void: () => {},
};

export function makeContractCreator<D extends GenericDescriptor>(_abi: any[], _descriptor: D) {
  return (web3: Web3, address: string): ContractWrapper<D> => {
    const baseContract = new web3.eth.Contract(_abi, address);

    const methodsProxy = new Proxy<Contract['methods']>(
      {},
      {
        get(target: Contract['methods'], prop: string) {
          const callMethodDescriptor = _descriptor.callMethods[prop];
          const sendMethodDescriptor = _descriptor.sendMethods[prop];

          if (!callMethodDescriptor && !sendMethodDescriptor) {
            return target[prop];
          }

          const { inputs: callInputs = [], output: callOutput } =
            callMethodDescriptor || sendMethodDescriptor;

          const baseCallFunction = (
            input: void | Record<string, JSType | JSType[]>,
            eventsForReload?: EventEmitter<any> | EventEmitter<any>[],
            updatingDelay?: number,
            tx?: CallOptions,
          ) => {
            return getContractData$(baseContract, web3.eth, prop, {
              tx,
              args: input
                ? callInputs.map(({ name, type }) => convertInputValueToRequest(type, input[name]))
                : [],
              eventsForReload,
              updatingDelay,
              convert: makeConvertFromResponse(callOutput && [...callOutput]),
            }) as Observable<any>;
          };

          const callFunction: CallMethod<any> = baseCallFunction;

          const readFunction: ReadMethod<any> = (
            input: void | Record<string, JSType | JSType[]>,
            tx: SendOptions,
            eventsForReload?: EventEmitter<any> | EventEmitter<any>[],
            updatingDelay?: number,
          ) => baseCallFunction(input, eventsForReload, updatingDelay, tx);

          if (callMethodDescriptor) {
            return callFunction;
          }

          if (sendMethodDescriptor) {
            const { inputs = [] } = sendMethodDescriptor;

            const getTransactionFunction = (
              input: void | Record<string, BN | string | boolean>,
            ) => {
              const args = input
                ? inputs.map(({ name, type }) => convertInputValueToRequest(type, input[name]))
                : [];
              return baseContract.methods[prop](...args);
            };

            const sendFunction: SendMethod<any> = attachStaticFields(
              (input: void | Record<string, BN | string | boolean>, tx: SendOptions) => {
                return getTransactionFunction(input).send(tx);
              },
              {
                getTransaction: getTransactionFunction,
                read: readFunction,
              },
            );

            return sendFunction;
          }

          return target[prop];
        },
      },
    );

    return (new Proxy<Contract>(baseContract, {
      get(target, prop: keyof Contract) {
        if (prop === 'methods') {
          return methodsProxy;
        }
        if (prop === 'events') {
          return target[prop];
        }
        return target[prop];
      },
    }) as unknown) as ContractWrapper<D>;
  };
}

export function attachStaticFields<T extends {}, I extends Record<string, any>>(
  target: T,
  staticFields: I,
): T & I {
  const result: T & I = target as T & I;

  Object.keys(staticFields).forEach((key: keyof I) => {
    (result as I)[key] = staticFields[key];
  });

  return result;
}

function convertInputValueToRequest(
  type: InputEvmType | Input<InputEvmType, boolean>,
  inputValue: JSType | JSType[],
): JSType | JSType[] {
  function convert(value: JSType | JSType[]) {
    return typeof type === 'string'
      ? (toRequest[type] as any)(value)
      : convertInputValueToRequest(type.type, value);
  }
  return Array.isArray(inputValue) ? inputValue.map(convert) : convert(inputValue);
}

function makeConvertFromResponse(output?: Output<OutputEvmType, boolean>[]) {
  return (value: JSType | JSType[]) => {
    if (!output || !output.length) {
      return value;
    }
    return output.length > 1
      ? output.map((outputItem, index) =>
          convertOutputValueFromResponse(outputItem.type, (value as JSType[])[index]),
        )
      : convertOutputValueFromResponse(output[0].type, value as JSType);
  };
}

function convertOutputValueFromResponse(
  type: OutputEvmType,
  value: JSType | JSType[],
): JSType | JSType[] {
  return Array.isArray(value)
    ? value.map(fromResponse[type] as any)
    : (fromResponse[type] as any)(value);
}

/* ***** MERGE ***** */

type MergeArguments<T, K extends string = 'whatever'> = {
  [Key in K]: T extends (first: infer A) => void ? A : MergeOnePlus<T, K>;
}[K];

type MergeOnePlus<T, K extends string> = {
  [Key in K]: T extends (first: infer A, ...args: infer U) => void
    ? A & MergeArguments<(...args: U) => void, K>
    : never;
}[K];

type IntoSignature<T extends readonly unknown[]> = (...args: T) => void;

type MergeTupleMembers<T extends readonly unknown[] | {}> = T extends readonly unknown[]
  ? MergeArguments<IntoSignature<T>>
  : never;
