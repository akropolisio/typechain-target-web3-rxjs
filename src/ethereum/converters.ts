import BN from 'bn.js';
import { RawAbiParameter } from 'typechain';

import {
  JSType,
  EvmType,
  Arguments,
  Response,
  InputEvmTypeToJSTypeMap,
  OutputEvmTypeToJSTypeMap,
  Web3ContractInputArgs,
  Web3ContractResponse,
  ArgumentType,
  isPlainValue,
} from './types';
import { attachStaticFields } from './utils';

export function inputsToRequest(
  inputsAbi: RawAbiParameter[],
  args: Arguments,
): Web3ContractInputArgs {
  return inputsAbi.map((abiParameter, index) => {
    const { isArray, type } = parseEvmType(abiParameter.type);
    const arg = (abiParameter.name && args[abiParameter.name]) || args[index];
    if (!arg) {
      throw new Error(`ABI parameter ${abiParameter.name} not found in ${args}`);
    }

    if (isArray) {
      if (!Array.isArray(arg)) {
        throw new Error(`Expected an array, but received ${arg}`);
      }
      return arg.map(innerArgs =>
        type === 'tuple'
          ? convertTupleValueToRequest(abiParameter, innerArgs)
          : convertPlainValueToRequest(type, innerArgs),
      );
    }

    return type === 'tuple'
      ? convertTupleValueToRequest(abiParameter, arg)
      : convertPlainValueToRequest(type, arg);
  });
}

function convertTupleValueToRequest(abiParameter: RawAbiParameter, args: ArgumentType) {
  if (!abiParameter.components || isPlainValue(args)) {
    throw new Error(`Unable to convert ${args} to a tuple request`);
  }
  return inputsToRequest(abiParameter.components, args as Arguments);
}

function convertPlainValueToRequest(type: EvmType, inputValue: ArgumentType): JSType {
  if (!isPlainValue(inputValue)) {
    throw new Error(`Expected string, boolean, or BN, but received ${inputValue}`);
  }

  const toRequest: {
    [key in EvmType]: (input: InputEvmTypeToJSTypeMap[key]) => JSType;
  } = {
    address: value => value,
    boolean: value => value,
    integer: value => value.toString(),
    uinteger: value => value.toString(),
    string: value => value,
    bytes: value => value,
    'dynamic-bytes': value => value,
  };

  return (toRequest[type] as any)(inputValue);
}

export function responseToOutput(
  outputsAbi: RawAbiParameter[],
  response: Web3ContractResponse,
): Response {
  if (outputsAbi.length === 1) {
    return convertResponseValueToOutput(outputsAbi[0], response);
  }

  return toMixedArray(outputsAbi, response);
}

function convertResponseValueToOutput(
  abiParameter: RawAbiParameter,
  responseValue: Web3ContractResponse,
): Response {
  const { type, isArray } = parseEvmType(abiParameter.type);

  if (isArray) {
    if (!Array.isArray(responseValue)) {
      throw new Error(`Expected an array, but received ${responseValue}`);
    }

    return responseValue.map(innerValues =>
      type === 'tuple'
        ? convertTupleResponseToOutput(abiParameter, innerValues)
        : convertPlainResponseToOutput(type, innerValues),
    );
  }
  return type === 'tuple'
    ? convertTupleResponseToOutput(abiParameter, responseValue)
    : convertPlainResponseToOutput(type, responseValue);
}

function convertTupleResponseToOutput(
  abiParameter: RawAbiParameter,
  responseValue: Web3ContractResponse,
) {
  if (!abiParameter.components) {
    throw new Error(`Components property is required for tuples`);
  }
  return toMixedArray(abiParameter.components, responseValue);
}

function toMixedArray(abiParameters: RawAbiParameter[], value: Web3ContractResponse) {
  if (isPlainValue(value)) {
    throw new Error(`Expected an object, or an array, but received ${value}`);
  }
  return abiParameters.reduce((resultAcc, componentAbi, index) => {
    const convertedValue = convertResponseValueToOutput(componentAbi, value[index]);
    resultAcc.push(convertedValue);
    if (componentAbi.name) {
      attachStaticFields(resultAcc, { [componentAbi.name]: convertedValue });
    }
    return resultAcc;
  }, [] as Response[]);
}

function convertPlainResponseToOutput(type: EvmType, value: Web3ContractResponse): JSType {
  if (!isPlainValue(value)) {
    throw new Error(`Expected string, boolean, or BN, but received ${value}`);
  }

  const toOutput: {
    [key in EvmType]: (typedValue: OutputEvmTypeToJSTypeMap[key]) => JSType;
  } = {
    address: v => v,
    boolean: v => v,
    integer: v => new BN(v),
    uinteger: v => new BN(v),
    string: v => v,
    bytes: v => v,
    'dynamic-bytes': v => v,
  };

  return (toOutput[type] as any)(value);
}

function parseEvmType(abiType: string): { type: EvmType | 'tuple'; isArray: boolean } {
  const [, inputType, isArray] = abiType.match(/^(.+?)(\[\d*?\])?$/) || [];

  const type = ((): EvmType | 'tuple' => {
    // eslint-disable-next-line default-case
    switch (inputType) {
      case 'bool':
        return 'boolean';
      case 'address':
        return 'address';
      case 'string':
        return 'string';
      case 'byte':
        return 'bytes';
      case 'bytes':
        return 'dynamic-bytes';
    }

    if (inputType.startsWith('tuple')) {
      return 'tuple';
    }
    if (inputType.startsWith('uint')) {
      return 'uinteger';
    }
    if (inputType.startsWith('int')) {
      return 'integer';
    }
    if (inputType.startsWith('bytes')) {
      return 'bytes';
    }

    throw new Error(`Type "${abiType}" cannot be parsed`);
  })();

  return { type, isArray: !!isArray };
}
