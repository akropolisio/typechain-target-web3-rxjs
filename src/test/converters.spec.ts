import BN from 'bn.js';

import { inputsToRequest, responseToOutput } from 'ethereum/converters';

describe('inputsToRequest', () => {
  it('should convert an object of arguments to an array of web3.js arguments', () => {
    const inputsAbi = [
      {
        name: 'a',
        type: 'tuple',
        components: [
          {
            name: 'x',
            type: 'tuple[]',
            components: [
              { name: 't', type: 'string[]' },
              { name: 'p', type: 'address[]' },
            ],
          },
          { name: 'y', type: 'int256[]' },
          { name: 'z', type: 'bool[]' },
        ],
      },
    ];
    const args = {
      a: {
        x: [
          { t: ['string', 'string'], p: ['0x00', '0x00'] },
          { t: ['string', 'string'], p: ['0x00', '0x00'] },
        ],
        y: [new BN('190000000000000000000'), new BN('210000000000000000000')],
        z: [true, false, true],
      },
    };
    expect({ args, result: inputsToRequest(inputsAbi, args) }).toMatchSnapshot();
  });

  it('should throw an error when args mismatch ABI', () => {
    const inputsAbi = [
      { name: 'a', type: 'uint256[]' },
      { name: 'b', type: 'address' },
    ];
    const args = { A: new BN(0) };
    const errorCall = () => inputsToRequest(inputsAbi, args);
    expect(errorCall).toThrowErrorMatchingSnapshot();
  });

  it('should support unnamed arguments', () => {
    const inputsAbi = [
      { name: '', type: 'uint256[]' },
      { name: 'b', type: 'address' },
    ];
    const args = {
      0: [new BN('190000000000000000000'), new BN('210000000000000000000')],
      b: '0x00',
    };
    expect({ args, result: inputsToRequest(inputsAbi, args) }).toMatchSnapshot();
  });
});

describe('responseToOutput', () => {
  it('should return array of tuple components when length equals 1', () => {
    const outputsAbi = [
      {
        name: 'a',
        type: 'tuple',
        components: [
          { name: 'x', type: 'uint256' },
          { name: 'y', type: 'address' },
          { name: 'z', type: 'bool' },
        ],
      },
    ];

    const response = ['190000000000000000000', '0x00', true];
    expect({
      response,
      result: responseToOutput(outputsAbi, response),
    }).toMatchSnapshot();
  });

  it('should return a mixed array when ABI length is greater than 1', () => {
    const outputsAbi = [
      { name: 'a', type: 'address' },
      {
        name: 'b',
        type: 'tuple[]',
        components: [
          {
            name: 'x',
            type: 'tuple',
            components: [
              {
                name: '',
                type: 'string',
              },
            ],
          },
          { name: 'y', type: 'bool[]' },
        ],
      },
    ];

    const response = {
      0: '0x00',
      1: [[['string'], [true, false]]],
      a: '0x00',
      b: [[['string'], [true, false]]],
    };
    expect({
      response,
      result: responseToOutput(outputsAbi, response),
    }).toMatchSnapshot();
  });

  it('should throw an error when response mismatches ABI', () => {
    const outputsAbi = [
      { name: '', type: 'uint256' },
      { name: '', type: 'address' },
    ] as any;
    const response = '190000000000000000000';
    const errorCall = () => responseToOutput(outputsAbi, response);
    expect(errorCall).toThrowErrorMatchingSnapshot();
  });

  it('should support unnamed parameters', () => {
    const outputsAbi = [
      { name: '', type: 'uint256' },
      { name: '', type: 'address' },
    ];
    const response = { 0: '190000000000000000000', 1: '0x00' };
    expect({
      response,
      result: responseToOutput(outputsAbi, response),
    }).toMatchSnapshot();
  });

  it('should support both numeric and string keys', () => {
    const outputsAbi = [
      { name: 'a', type: 'uint256' },
      { name: 'b', type: 'address[]' },
      {
        name: 't',
        type: 'tuple[]',
        components: [
          { name: 'x', type: 'string' },
          { name: 'y', type: 'address[]' },
          { name: 'z', type: 'bool' },
        ],
      },
    ];
    const response = {
      0: '190000000000000000000',
      1: ['0x00', '0x00'],
      2: [
        ['Hello', ['0x00', '0x00', '0x00'], true],
        ['Hello', ['0x00', '0x00', '0x00'], false],
      ],
      a: '190000000000000000000',
      b: ['0x00', '0x00'],
      t: [
        ['Hello', ['0x00', '0x00', '0x00'], true],
        ['Hello', ['0x00', '0x00', '0x00'], false],
      ],
    };
    const output = responseToOutput(outputsAbi, response);
    expect({ response, tupleComponentKeys: Object.keys((output as any).t[0]) }).toMatchSnapshot();
  });
});
