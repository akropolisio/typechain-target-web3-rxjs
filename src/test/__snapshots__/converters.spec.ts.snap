// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inputsToRequest should convert an object of arguments to an array of web3.js arguments 1`] = `
Object {
  "args": Object {
    "a": Object {
      "x": Array [
        Object {
          "p": Array [
            "0x00",
            "0x00",
          ],
          "t": Array [
            "string",
            "string",
          ],
        },
        Object {
          "p": Array [
            "0x00",
            "0x00",
          ],
          "t": Array [
            "string",
            "string",
          ],
        },
      ],
      "y": Array [
        "0a4cc799563c380000",
        "0b6255df5f50080000",
      ],
      "z": Array [
        true,
        false,
        true,
      ],
    },
  },
  "result": Array [
    Array [
      Array [
        Array [
          Array [
            "string",
            "string",
          ],
          Array [
            "0x00",
            "0x00",
          ],
        ],
        Array [
          Array [
            "string",
            "string",
          ],
          Array [
            "0x00",
            "0x00",
          ],
        ],
      ],
      Array [
        "190000000000000000000",
        "210000000000000000000",
      ],
      Array [
        true,
        false,
        true,
      ],
    ],
  ],
}
`;

exports[`inputsToRequest should support unnamed arguments 1`] = `
Object {
  "args": Object {
    "0": Array [
      "0a4cc799563c380000",
      "0b6255df5f50080000",
    ],
    "b": "0x00",
  },
  "result": Array [
    Array [
      "190000000000000000000",
      "210000000000000000000",
    ],
    "0x00",
  ],
}
`;

exports[`inputsToRequest should throw an error when args mismatch ABI 1`] = `"ABI parameter a not found in [object Object]"`;

exports[`responseToOutput should return a mixed array when ABI length is greater than 1 1`] = `
Object {
  "response": Object {
    "0": "0x00",
    "1": Array [
      Array [
        Array [
          "string",
        ],
        Array [
          true,
          false,
        ],
      ],
    ],
    "a": "0x00",
    "b": Array [
      Array [
        Array [
          "string",
        ],
        Array [
          true,
          false,
        ],
      ],
    ],
  },
  "result": Array [
    "0x00",
    Array [
      Array [
        Array [
          "string",
        ],
        Array [
          true,
          false,
        ],
      ],
    ],
  ],
}
`;

exports[`responseToOutput should return array of tuple components when length equals 1 1`] = `
Object {
  "response": Array [
    "190000000000000000000",
    "0x00",
    true,
  ],
  "result": Array [
    "0a4cc799563c380000",
    "0x00",
    true,
  ],
}
`;

exports[`responseToOutput should support both numeric and string keys 1`] = `
Object {
  "response": Object {
    "0": "190000000000000000000",
    "1": Array [
      "0x00",
      "0x00",
    ],
    "2": Array [
      Array [
        "Hello",
        Array [
          "0x00",
          "0x00",
          "0x00",
        ],
        true,
      ],
      Array [
        "Hello",
        Array [
          "0x00",
          "0x00",
          "0x00",
        ],
        false,
      ],
    ],
    "a": "190000000000000000000",
    "b": Array [
      "0x00",
      "0x00",
    ],
    "t": Array [
      Array [
        "Hello",
        Array [
          "0x00",
          "0x00",
          "0x00",
        ],
        true,
      ],
      Array [
        "Hello",
        Array [
          "0x00",
          "0x00",
          "0x00",
        ],
        false,
      ],
    ],
  },
  "tupleComponentKeys": Array [
    "0",
    "1",
    "2",
    "x",
    "y",
    "z",
  ],
}
`;

exports[`responseToOutput should support unnamed parameters 1`] = `
Object {
  "response": Object {
    "0": "190000000000000000000",
    "1": "0x00",
  },
  "result": Array [
    "0a4cc799563c380000",
    "0x00",
  ],
}
`;

exports[`responseToOutput should throw an error when response mismatches ABI 1`] = `"Expected an object, or an array, but received 190000000000000000000"`;
